torch/csrc/cuda/Storage.cpp

```diff
@@ -18,8 +18,22 @@
 #define THC_GENERIC_FILE "torch/csrc/generic/Storage.cpp"
 #include <THC/THCGenerateAllTypes.h>
 
+#ifndef USE_CUDA
+#error "Compiling torch/csrc/cuda/Storage.cpp without USE_CUDA"
+#endif
+
+// NB: When !USE_CUDA, the implementation of this lives
+// in torch/csrc/Storage.cpp.
+// If you ever divest libtorch of USE_CUDA, you'll have to virtualize
+// the CUDA call.
 template<>
-void THPPointer<THCStorage>::free() {
-  if (ptr)
-    THCStorage_free(LIBRARY_STATE ptr);
+void THPPointer<THStorage>::free() {
+  if (ptr) {
+    if (ptr->backend == at::kCPU) {
```

Comment:
can you just have a free function that takes in backend and dispatches correctly?

Reply:
Not easily, because I can't put the function in TH/THC, since it needs to cross the CPU to CUDA gap, even when CPU and CUDA implementations are in separate shared library. The gap-crossing isn't necessary once we have a single implementation of free, but that's blocked on the allocator. I agree the intent is for this to get killed when we do have a unified free function on storage, but it seems not worth adding a CUDAHook just for this in its current form.

Reply:
I'm not sure I buy this -- isn't there a bunch of code in torch/csrc that is ifdef'ed or is this significantly different?  (I agree it's not worth putting in CUDAHook).

Reply:
Well, the key is where the code lives. If the free function is in TH/THC, we don't have `USE_CUDA` macro (because we have done it the "clean" way inside ATen, with a separate library for CUDA/non-CUDA, no preprocessor macros). Inside `torch/csrc`, we haven't done it "cleanly" yet, so you can ifdef it.